src/librustc_typeck/check/op.rs

```diff
@@ -246,76 +246,146 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                                hir::BiLt | hir::BiLe | hir::BiGt | hir::BiGe =>
+                                    Some("std::cmp::PartialOrd"),
+                                _             => None
+                            };
+                            let mut suggested_deref = false;
+                            if let TyRef(_, ref ty_mut) = lhs_ty.sty {
+                                if {
+                                    !self.infcx.type_moves_by_default(self.param_env,
+                                                                        ty_mut.ty,
+                                                                        lhs_expr.span) &&
+                                        self.lookup_op_method(ty_mut.ty,
+                                                                &[rhs_ty],
+                                                                Op::Binary(op, is_assign))
+                                            .is_ok()
+                                } {
+                                    let codemap = self.tcx.sess.codemap();
+                                    match codemap.span_to_snippet(lhs_expr.span) {
+                                        Ok(lstring) =>{
+                                            let msg = &format!(
+                                                "`{}` can be used on '{}', you can \
+                                                dereference `{2}`: `*{2}`",
+                                                op.node.as_str(), ty_mut.ty, lstring);
+                                            err.help(msg);
+                                            suggested_deref = true;
+                                        },
+                                        _ =>{}
+                                    }
+                                }
                             }
+                            if let Some(missing_trait) = missing_trait {
+                                if missing_trait == "std::ops::Add" &&
+                                    self.check_str_addition(expr, lhs_expr, rhs_expr, lhs_ty,
+                                                            rhs_ty, &mut err) {
+                                    // This has nothing here because it means we did string
+                                    // concatenation (e.g. "Hello " + "World!"). This means
+                                    // we don't want the note in the else clause to be emitted
+                                } else if let ty::TyParam(_) = lhs_ty.sty {
+                                    // FIXME: point to span of param
+                                    err.note(
+                                        &format!("`{}` might need a bound for `{}`",
+                                                    lhs_ty, missing_trait));
+                                } else {
+                                    if !suggested_deref{
```

Comment:
[nitpick] missing space

This can be `} else if !suggested_deref {` without the need of an extra indentation level.